### ISSUE TITLE
Better handling of uploading and inserting media files in new posts on Friendica

### ISSUE THREAD

**Issue by ghost** [friendica/friendica#8674]
Compared to other platforms, incorporating media into Friendica is, to put it mildly, rather rudimentary.

For photos, after you have uploaded them, you have to select them again from the gallery.
Wouldn't it be more elegant if the photo, after uploading, would appear directly in the composer?

With videos, there are additional steps compared to the photo upload.
In order for the video upload to work, you also have to switch to the Files tab in the upload window. Only then, can you upload a video file, which must then be selected again with an additional click.

Wouldn't it be more elegant if Friendica, after uploading the file, automatically recognizes it and inserts it into the composer, whether video or image file?

**Comment by MrPetovan** [comment]
Yes and no. We have a complex media/file library with folders and subfolders and such a default behavior would make it irrelevant which isn't great either.

**Comment by copiis** [comment]
It would be a step forward if it were possible to upload videos via the mobile apps.

**Comment by MrPetovan** [comment]
Indeed, not sure what's needed if it just is a missing API call or a better uploading mechanism on mobile web?

**Comment by Extarys** [comment]
>Yes and no. We have a complex media/file library with folders and such a default behavior would make it irrelevant which isn't great either. 

@MrPetovan Once a photo is uploaded, is it possible to change folder, etc? If so, a simple generic folder called "Uploads" could be used.

It's been a while I can't remember how to handle uploads in PHP. I think you need a special page that is called every second or every 500ms that would return to the client the progress of the upload.

I will need help for the server side of things though. This is actually a reoccurring issue I find here so I might do this soon.

**Comment by MrPetovan** [comment]
I don't even know if it's possible to change folder right now, but I'm pretty sure it could be done without breaking existing URLs.

**Comment by Extarys** [comment]
I believe facebook have a folder called Timeline or Wall or something but it's been a long time since I uploaded pictures there.

Thinking of it, facebook for example allows user to move pictures around, you can create albums from the other pictures - I might be wrong here.

You can also upload a group of pictures and assign a folder,  then fb can create a post containing those pictures. (Again I dont know if it'S only the new one or if it creates a post will like the new one first but all the pictures browsable from the get go.

Since friendica is federated, this will require some thinking.

**Comment by AlfredSK** [comment]
We already use a default folder when uploading a photo directly from the post composer/editor. The problem is Friendica is switching the folder name when you change your language setting: EN -> 'Wall Photos'; DE -> 'Pinnwand-Bilder'; ...
The folder name is just a string in the database. You can move photos by changing the folder name in the database w/o losing photos or breaking links. I've already done that.
It is possible to change the folder a photo is associated with in the user interface by editing the photo. Go to your photos. Click on a photo. Click on the little pencil to start editing. There is a field for the album name. Change it. Save the photo.

**Comment by Extarys** [comment]
TLDR:
@MrPetovan will need to chip in as he's the one deciding and knows way more than me about friendica :laughing: 

Changing folder wouldn't/shouldn't change the url/image id



> The problem is Friendica is switching the folder name when you change your language setting: EN -> 'Wall Photos'; DE -> 'Pinnwand-Bilder'; ...

Well this is just text being displayed. `/photos/wallphotos` shouldn't need translation, it's an url. (EDIT: the url seems to only contain the photo id anyway)

I uploaded my first picture in a post. The  URL is `http://friendica.local/photos/testadmin/image/140220998960044bd46fee1000795101`

If I use the dropdown top-right and click Photos, I get 3 pictures. 2 of them says "Wall Photos" and if I do like you say I find the "Edit album name" field:
![image](https://user-images.githubusercontent.com/11581433/104846366-24cb8600-58a8-11eb-8e7c-bd7392fa31a9.png)

> It is possible to change the folder a photo is associated with in the user interface

But I didn't find a way to browse an album or add pictures to it. I can view a single picture when I click on it.

![image](https://user-images.githubusercontent.com/11581433/104846398-56445180-58a8-11eb-8cc8-58700e531117.png)


Anyway, since the url doesn't seem tied to an album as it's just a string, moving them around and defaulting post uploads to a Wall Photos folder seems to be the best solution.

I presume each album have a unique id? Then the album id for wall photos could be the same as the user's id. If this doesn't work because server/db upgrades and id collisions then **adding a _purpose_ field to each non-user created album could be the solution**. But I believe it would add size to all non-system albums.

Another possibility would be to add a table to track system albums. For example `sysalbums` with the userid, the albumid and a purpose. 0 for wall photos as the default. This method would be entirely transparent to the end user of course and on the API.

As an example the object can say:
**/v1/getAlbums/{userid}**
Get all albums from a user
```
{
    count: 2,
    albums: [
        {
            album_id: 997, <- fetched from sysalbum
            album_name: null, <- wall photos album is translated from the template or the client, not a custom name
            user_id: 1564,
            username: "testadmin"
            count: 16,
            special: "_WALLPHOTOS" <- this allows the client/template to know it's a default album and states it's purpose. PHP or any other then knows to show "Pinnwand-Bilder" if needed
        }
    ]
}
```

**/v1/getAlbums/{userid}/997**
Get all photos of an album from a user
```
user: testadmin
imgid: 140220998960044bd46fee1000795101
albumid: 997,
upload_date: ...
image_size: ...
```

I didn't check out the API much yet and this might interfere with older friendica version. It's to demonstrate that we can have default albums without affecting the name/translation of it.

**Comment by AlfredSK** [comment]
> I presume each album have a unique id?

As far as I know: no. Every photo has a record in the `photo` table. The album name is just a string (exactly as seen in the GUI) stored in the field `album` of type varchar. It's really just the name. That's it. Nothing else. There is no abstraction, no album id.
It's a very simple (or some may say humble) solution to create/organize the "albums".

Disclaimer: The DB structure is changing constantly. So it is possible that my knowledge is outdated. :-)

**Comment by Extarys** [comment]
Thanks for clarifying. I saw some work being done on the database side, hopefully this will also get a rework eventually.

Then yes, in this specific case, the only solution I can find is to use some kind of string as the album name to tell the client it's a wall photo. Then again if the end user name their album with the special string... :man_shrugging:

**Comment by foss-** [comment]
I know some friendica users have come to love the gallery. I personally have no use for it and assume the majority of users would just want a simple photo upload (on desktop: drag and drop image to new post, on mobile: click image gallery button and select image from phone photo roll or camera) and then forget about the image once the post is out. Can't we have both? The current workflow is cumbersome to the point that I don't bother anymore to create posts with images on friendica.

I believe the friendica gallery / or rather current workflow does more harm than it helps users.

**Comment by MrPetovan** [comment]
You are right, but at least the current image insertion workflow exists. Anything else would require additional development efforts.

**Comment by foss-** [comment]
Image upload has been re-written and is now much more straigh-forward and allows dropping images to a field below the post composer which will then auto-add the image to the post:

<img src="https://github.com/friendica/friendica/assets/5708172/05549a50-f0b3-4d35-a38d-adb6eec372ef" width="450">

**Comment by MrPetovan** [comment]
Doesn’t it satisfy the original ask?

**Comment by foss-** [comment]
Follow-up issue for the missing video upload feature: https://github.com/friendica/friendica/issues/13992
